tests/Functional/Controller/Admin/GroupControllerTest.php

```diff
@@ -42,55 +48,140 @@ protected function tearDown(): void
+        $id = $group->getId();
+
+        // Act
+        $crawler = $this->client->request('GET', $this->controllerEndpoint . "/{$id}/delete");
+        $form = $crawler->selectButton('Ja, verwijder')->form();
+        $crawler = $this->client->submit($form);
+
+        $allGroups = $this->em->getRepository(Group::class)->findAll();
+
+        // Assert
+        $this->assertEquals(200, $this->client->getResponse()->getStatusCode());
+        $this->assertNotContains($group, $allGroups);
     }
 
     public function testRelationNewAction(): void
     {
-        /* @todo This test is incomplete. */
-        $this->markTestIncomplete();
+        // Arrange
+        $group = $this->em->getRepository(Group::class)->findAll()[0];
+        $id = $group->getId();
+
+        // Act
+        $crawler = $this->client->request('GET', $this->controllerEndpoint . "/relation/new/{$id}");
+        $form = $crawler->selectButton('Toevoegen')->form();
+        $crawler = $this->client->submit($form);
+        $allGroups = $this->em->getRepository(Group::class)->findAll();
+        $newGroupId = explode('group/', $crawler->getUri())[1];
+        $newGroup = $this->em->getRepository(Group::class)->find($newGroupId);
+
+        // Assert
+        $this->assertEquals(200, $this->client->getResponse()->getStatusCode());
+        $this->assertContains($newGroup, $allGroups);
     }
 
     public function testRelationAddAction(): void
     {
-        /* @todo This test is incomplete. */
-        $this->markTestIncomplete();
+        // Arrange
+        $relation = $this->em->getRepository(Relation::class)->findAll()[0];
+        $id = $relation->getId();
+
+        // Act
+        $crawler = $this->client->request('GET', $this->controllerEndpoint . "/relation/add/{$id}");
+        $form = $crawler->selectButton('Toevoegen')->form();
+        $form['relation_add[_token]'] = "Penningmeester";
+        $crawler = $this->client->submit($form);
+        $allRelations = $this->em->getRepository(Relation::class)->findAll();
+        $newGroupId = explode('add/', $crawler->getUri())[1];
+        $newRelation = $this->em->getRepository(Relation::class)->find($newGroupId);
+
+        // Assert
+        $this->assertEquals(200, $this->client->getResponse()->getStatusCode());
+        $this->assertContains($newRelation, $allRelations);
     }
+    
 
     public function testRelationDeleteAction(): void
     {
-        /* @todo This test is incomplete. */
-        $this->markTestIncomplete();
+        // Arrange
+        $relation = $this->em->getRepository(Relation::class)->findAll()[0];
+        $id = $relation->getId();
+
+        // Act
+        $crawler = $this->client->request('GET', $this->controllerEndpoint . "/relation/delete/{$id}");
+        $form = $crawler->selectButton('Ja, verwijder')->form();
+        $crawler = $this->client->submit($form);
+
+        $allRelations = $this->em->getRepository(Relation::class)->findAll();
```

Comment:
I feel like this line can be moved to the `// Arrange` block, eg.
```php
$allRelations = $this->em->getRepository(Relation::class)->findAll();
$relation = $allRelations[0];
$id = $relation->getId();```

Reply:
I see what you mean, however I think you misunderstood the flow of the test.

`$allRelations` fetches all the relations after the deletion event to check it does not longer contain the original relation. Thus confirming the deletion was successful.